### PR TITLE
 Order Fulfillment: Allow the user to search and filter shipment providers

### PR DESCRIPTION
**Pull request by ctarda**
Implements #850 

This PR adds functionality to filter the list of shipment providers.

## Screenshots
All screenshots and recordings are on the iPhone 8 simulator
### Full List
Simulator Screen Shot - iPhone 8 - 2019-04-12 at 14 00 32

### Filtered list
Simulator Screen Shot - iPhone 8 - 2019-04-12 at 14 00 36

### No results
Simulator Screen Shot - iPhone 8 - 2019-04-12 at 14 10 21

### In action
Screen Recording 2019-04-12 at 2 00 16 PM mov

## Diff
- Changed the way the `resultsController` is constructed in `ShippingProvidersViewModel`. I did not rename this class to `ShipmentProvidersViewModel` so that the diff could be isolated in the review. If / when this PR is ready to merge, we can rename this class.
- Renamed `ShippingProvidersViewController` to `ShipmentProvidersViewController`, added back the search bar, and pushed the result of users input into the view model so that the view model can filter the date
- Added some styles and setup the appearance of search bars
- Added some tests

## Testing
- Checkout the branch, run the tests
- Add a breakpoint in line 271 of `ShipmentProvidersViewController` (method `addCustomProvider`)
- Navigate to an order > Fulfill order > Add tracking > Shipping Provider
- Filter the list (i.e. typing DHL). Notice only providers matching the search term appear in the list
- Filter by a non-existent provider (for example `Zz`). That should present an "empty state" screen. 
- Tap the "Custom Provider" button, execution should stop at the breakpoint added previously.

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/ShipmentProvidersViewController.swift [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        guard viewModel.isListEmpty else {
+            removeEmptyState()
+            return
+        }
+
+        let emptyState: EmptyListMessageWithActionView = EmptyListMessageWithActionView.instantiateFromNib()
+        emptyState.messageText = NSLocalizedString("No results found for \(term)\nAdd a custom provider",
+            comment: "Empty state for the list of shipment providers. It reads: 'No results for DHL. Add a custom provider'")
+        emptyState.actionText = NSLocalizedString("Custom Provider",
+                                                  comment: """
```
[the remark this answers is not in the document]
Comment (ctarda) [reply WooCommerce/Classes/ViewRelated/Orders/ShipmentProvidersViewController.swift]: Right! It was mentioned a few weeks back, sorry about that!

**Comment by ctarda** [comment]
Thanks for taking a look @mindgraffiti I just pushed an update to remove the triple-quotes

**Comment by mindgraffiti** [comment]
Looks good! 

:shipit:

File: WooCommerce/Classes/Extensions/UISearchBar+Appearance.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        let textAttributes = [NSAttributedString.Key.foregroundColor: StyleManager.wooGreyMid]
+        UITextField.appearance(whenContainedInInstancesOf: [UISearchBar.self]).defaultTextAttributes = textAttributes
+    }
+}
+
+
+// Setting the UITextField background color via the appearance proxy does not seem to work
+// As a workaround, this property exposes the texfield, so the background color can be set manually
+//
+extension UISearchBar {
```
[the remark this answers is not in the document]
Comment (ctarda) [reply WooCommerce/Classes/Extensions/UISearchBar+Appearance.swift]: It is indeed... :trollface:

**Pull request merged**